### PR TITLE
[pkg/collector] Add worker utilization stats to expvars

### PR DESCRIPTION
**Pull request by sgnn7**
### What does this PR do?

This PR:
- Adds periodic (every 15s) collection of utilization stats (how much time was spent running checks) for each worker.
- Calculates an average over 5 minutes or however long the worker was running, whichever is less
- Updates the `expvars["runner"]["Workers"]["Instances"]["instance_name"].Utilization` expvar after each new data point
- Changes `expvars["runner"]["Workers"]` expvar that contained the worker count to `expvars["runner"]["Workers"]["Count"]`
- ~Shows a warning in log if any worker that is not running a long check (e.g. jmxfetch) utilization goes over a predefined utilization threshold (`0.99`)~ (removed pending follow-up task)

### Motivation

It's not easy to track down problems with worker utilization just from logs themselves. This change should allow much quicker debugging of busy agents.

### Additional Notes

Anything else we should know when reviewing?

#### Basic functionality

- Set some number of `check_runners` in config
- Run the agent
- Initially the worker stats (`curl -s localhost:5000/debug/vars | jq ".runner.Workers" `) show correctly:
   - Names of each runner and the utilization is `0`
   - Correct count of the runners
- After 15s and over time, the value will stabilize and ensure:
   - Names of each runner and the utilization (range: 0 < util < 1)
   - Correct count of the runners
- Stop the agent
- Change the `check_runners` to 1 and start the agent
-  Initially the worker stats (`curl -s localhost:5000/debug/vars | jq ".runner.Workers" `) show correctly:
   - Correct count (`1`) of the runners
   - Name of the runner (`worker_*`) and the utilization is `0`
- After 15s and over time, the value will stabilize and ensure:
   - Correct count of the runners
   - Name of the runner (`worker_*`) 
   - Utilization proportional to the multiple of the original test scenario (e.g.`0.02` avg per worker w/ `check_runners = 4` should be close to `0.08` utilization for 1 worker with a max of `1.00`)
   
#### Basic functionality pt2

- Using the values collected before, add more checks (or make them run longer) and observe the increase in utilization
- Using the values collected before, remove some checks (or make them run shorter) and observe the decrease in utilization
- Using the values collected before, disable/remove all checks and observe the utilization stabilize at 0 after 5 min
- A long-check (eg. jmxfetch check) should have a worker stuck at 1.0 utilization (as it's keeping the child process running).

### Checklist

- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

**Comment by sgnn7** [comment]
Moving milestone to 7.32.0 since it look unlikely that this will get in before the freeze.

File: pkg/collector/worker/utilization_tracker.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+func (ut *utilizationTracker) Start() error {
+	ut.Lock()
+	defer ut.Unlock()
+
+	if ut.started {
+		return fmt.Errorf("Attempted to use UtilizationTracker.Start() when the tracker was already started")
+	}
+
+	if ut.stopped {
+		return fmt.Errorf("Attempted to use UtilizationTracker.Start() when the tracker after it was stopped")
```
Comment (djmitche) [inline pkg/collector/worker/utilization_tracker.go:127]:
```suggestion
		return fmt.Errorf("Attempted to use UtilizationTracker.Start() after the tracker was stopped")
```

File: pkg/util/sliding_window.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// PollingFunc is polled at specified intervals to collect the value
+// that should be included in the slidingWindow data
+type PollingFunc func() (value float64)
+
+// CallbackFunc is invoked when the polling function collects data and
+// we need to use the newly-calculated values
+type CallbackFunc func(SlidingWindow)
```
Comment (djmitche) [inline pkg/util/sliding_window.go:22]: This comment is pretty confusing.  Is it called when the stats are updated?  Maybe `StatsUpdatedFunc`?

File: pkg/util/sliding_window.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	windowSize      time.Duration
+}
+
+// SlidingWindow is the public API that we expose from the slidingWindow object
+type SlidingWindow interface {
+	Start(PollingFunc, CallbackFunc) error
+	Stop()
+
+	Average() float64
+	WindowSize() time.Duration
```
Comment (djmitche) [inline pkg/util/sliding_window.go:51]: It'd be good to have a doc comment on each method in the interface (in any interface, I think..)

File: pkg/util/sliding_window.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		pollingInterval: pollingInterval,
+		windowSize:      windowSize,
+	}, nil
+}
+
+// Start creates a new sliding window object, validates the parameters,
+// and starts the ticker. We use `Start` to define most of the variables
+// instead of the constructor as we are likely to need the SlidingWindow
+// instance in the polling/callback closures, leading to a chicken/egg
+// problem in usage.
```
Comment (djmitche) [inline pkg/util/sliding_window.go:82]: For `callbackFunc` the SlidingWindow is passed to the callback.  If the same were done for `pollingFunc`, would this issue be eliminated?

File: pkg/util/sliding_window.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+				sw.bucketsLock.Lock()
+
+				sw.buckets[sw.bucketIdx] = value
+				if sw.numBucketsUsed < sw.numBuckets {
+					sw.numBucketsUsed++
+				}
+
+				sw.bucketsLock.Unlock()
+
+				sw.bucketIdx = (sw.bucketIdx + 1) % sw.numBuckets
```
Comment (djmitche) [inline pkg/util/sliding_window.go:138]: This should probably be covered by bucketsLock?

File: pkg/util/sliding_window.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+type CallbackFunc func(SlidingWindow)
+
+// slidingWindow is an object that polls for a value every `pollingInterval`
+// and then keeps those values for the duration of the `windowSize` that can
+// then be analyzed/averaged/aggregated. This object should be instantiated
+// via `NewSlidingWindow`.
+type slidingWindow struct {
+	bucketIdx       int
+	buckets         []float64
+	bucketsLock     sync.RWMutex
```
Comment (djmitche) [inline pkg/util/sliding_window.go:31]: what fields does this protect (similarly for windowLock)?

File: pkg/util/sliding_window.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+	}
+
+	return sw.windowSize
+}
+
+// Average returns an average of all the polled values collected over the
+// sliding window range.
+func (sw *slidingWindow) Average() float64 {
+	if !sw.isInitialized() {
+		log.Warnf("Attempting to use SlidingWindow.Average() without initializting it!")
```
Comment (djmitche) [inline pkg/util/sliding_window.go:196]:
```suggestion
		log.Warnf("Attempting to use SlidingWindow.Average() without initializing it!")
```

**Review by djmitche — COMMENTED** [review]
A few comments below.  This generally looks good!

**Review by djmitche — APPROVED** [review]
Looks good! 👍🏽

**Pull request merged**